### PR TITLE
Fix the test_valgrind.sh script running tests under valgrind

### PR DESCRIPTION
**Pull request by ldorau**
The current script just checks if there is the
"ERROR SUMMARY: 0 errors from 0 contexts"
line in the output of a test.

It turned out that there can be many lines starting with
"ERROR SUMMARY:" in the output of one test - some with
errors and some without, so the current approach is wrong.
We have to check if there are no other lines starting with
"ERROR SUMMARY:" than this:
"ERROR SUMMARY: 0 errors from 0 contexts" instead.

### CODE REVIEW

**Comment by igchor** [comment]
@ldorau, the failures for `provider_os_memory_config` are expected since we are triggering segfaults intentionally. I think we need some kind of blacklist/filter for tests under valgrind.

**Comment by igchor** [comment]
But there are still some failures in `numa_modes` test (after commenting out  `protection_flag_*` tests) that are not reported after your fixes:

```
./test/umf_test-provider_os_memory_config (valgrind failed) - OK
./test/umf_test-scalable_pool - OK
PASSED
```

The reason is that the test fails under valgrind - but the failure is not a memory leak, we just fail on an assert.

**Comment by ldorau** [comment]
> @ldorau, the failures for `protection_flag_*` testcases in provider_os_memory_config` are expected since we are triggering segfaults intentionally. I think we need some kind of blacklist/filter for tests under valgrind.

I will try to add suppressions for them.

**Comment by ldorau** [comment]
> > @ldorau, the failures for `protection_flag_*` testcases in provider_os_memory_config` are expected since we are triggering segfaults intentionally. I think we need some kind of blacklist/filter for tests under valgrind.
> 
> I will try to add suppressions for them.

@igchor Suppressions added.

**Comment by igchor** [comment]
> > > @ldorau, the failures for `protection_flag_*` testcases in provider_os_memory_config` are expected since we are triggering segfaults intentionally. I think we need some kind of blacklist/filter for tests under valgrind.
> > 
> > 
> > I will try to add suppressions for them.
> 
> @igchor Suppressions added.

And what about the failures with get_mempolicy? In the logs, I still see: 
`./test/umf_test-provider_os_memory_config (valgrind failed) - OK` 

Are we just going to ignore this? I don't really know why this fails.

**Comment by ldorau** [comment]
> And what about the failures with get_mempolicy? In the logs, I still see: `./test/umf_test-provider_os_memory_config (valgrind failed) - OK`
> 
> Are we just going to ignore this? I don't really know why get_mempolicy does not work under valgrind.

Yes, I would ignore that. The most important is that this test passes without valgrind. Apparently valgrind changes behavior of this test. We cannot help this.

**Comment by ldorau** [comment]
The valgrind (Nightly) build passes: https://github.com/oneapi-src/unified-memory-framework/actions/runs/7731094933/job/21077971238?pr=189

**Comment by ldorau** [comment]
The corresponding Nightly CI job: https://github.com/ldorau/unified-memory-framework/actions/runs/7738886213

**Comment by DamianDuy** [comment]
LGTM

File: test/umf_test-provider_os_memory_config.supp
Comment (lukaszstolarczuk) [inline test/umf_test-provider_os_memory_config.supp:1]: perhaps, please make a separate sub-dir for suppressions

File: .github/workflows/nightly.yml [hunk 1]
```diff
@@ -41,5 +41,5 @@ jobs:
             cmake --build ${{github.workspace}}/build --config Debug -j$(nproc)
 
         - name: Run tests under valgrind
-          working-directory: ${{github.workspace}}/build
-          run: ../test/test_valgrind.sh ${{matrix.tool}}
+          working-directory: ${{github.workspace}}/build/test
```
Comment (lukaszstolarczuk) [inline .github/workflows/nightly.yml:44]: this line seems redundant now

File: test/test_valgrind.sh [hunk 2]
```diff
@@ -5,27 +5,35 @@
...
-if [ "$TOOL" == "" ]; then
-	echo "error: valgrind tool is missing"
+if [ ! -f $WORKSPACE/README.md ]; then
+	echo -e "error: incorrect <workspace_dir>: $WORKSPACE\n"
+	print_usage
+	exit 1
+fi
+
+if [ $(ls -1 ${BUILD_DIR}/test/umf_test-* 2>/dev/null | wc -l) -eq 0 ]; then
+	echo -e "error: UMF tests ${BUILD_DIR}/test/umf_test-* not found (maybe incorrect build directory: ${BUILD_DIR})\n"
```
Comment (lukaszstolarczuk) [inline test/test_valgrind.sh:36]: since `${BUILD_DIR}` is part of the path you're printing, printing it once more at the end seems redundant.

Comment (lukaszstolarczuk) [inline test/test_valgrind.sh:36]: you could perhaps suggest instead "(maybe tests not built)" 😉

File: test/test_valgrind.sh [hunk 3]
```diff
@@ -41,31 +49,45 @@ helgrind)
...
 echo "Running: \"valgrind $OPTION\" for the following tests:"
 
-for tf in $(ls -1 ./test/umf_test-*); do
-	[ ! -x $tf ] && continue
-	echo -n "$tf "
-	LOG=${tf}.log
-	HWLOC_CPUID_PATH=./cpuid valgrind $OPTION $tf >$LOG 2>&1 || echo -n "(valgrind failed) "
-	if grep -q -e "ERROR SUMMARY: 0 errors from 0 contexts" $LOG; then
+FAIL=0
+rm -f *.log *.err
```
Comment (lukaszstolarczuk) [inline test/test_valgrind.sh:72]: I know it's just a build dir, but perhaps be more precise and only remove your logs, meaning `rm umf_test*.log umf_test*.err`

File: test/umf_test-provider_os_memory_config.supp
Comment (ldorau) [reply test/umf_test-provider_os_memory_config.supp:1]: Done

File: .github/workflows/nightly.yml (hunk 1 above)
Comment (ldorau) [reply .github/workflows/nightly.yml:44]: Removed

File: test/test_valgrind.sh (hunk 2 above)
Comment (ldorau) [reply test/test_valgrind.sh:36]: Done

Comment (ldorau) [reply test/test_valgrind.sh:36]: Done

File: test/test_valgrind.sh (hunk 3 above)
Comment (ldorau) [reply test/test_valgrind.sh:72]: Done

**Review by lukaszstolarczuk — APPROVED** [review]

**Pull request merged**